src/libs/getArticleWithParentSections.js

```diff
@@ -1,48 +1,89 @@
 // @ts-check
+/**
+ * @typedef AgreementArticleWithParent
+ * @property {"article"} type
+ * @property {KaliData.AgreementArticleData} data
+ * @property {AgreementSectionWithParent} parent
+ *
+ * @typedef AgreementSectionWithParent
+ * @property {"section"} type
+ * @property {KaliData.AgreementSectionData} data
+ * @property {KaliData.AgreementArticle[]} children
+ * @property {AgreementSectionWithParent} parent
+ */
 
-const getAgreement = require("./getAgreement");
-const getArticles = require("./getArticles");
 const unistUtilFind = require("unist-util-find");
 
-const INDEXED_ARTICLES = getArticles();
+const getAgreement = require("./getAgreement");
+const getIndexedArticle = require("./getIndexedArticle");
 
 /**
- * Get an agreement article unist node with its parent sections.
+ * @param {KaliData.Agreement} agreement
+ * @param {Partial<KaliData.AgreementArticleData>} data
  *
- * @param {string} articleId
- *
- * @returns {KaliData.AgreementArticleWithParentSections}
+ * @returns {AgreementArticleWithParent=}
+ */
+const findArticleWithData = (agreement, data) =>
+  unistUtilFind(agreement, {
+    data,
+    type: "article",
+  });
+
+/**
+ * @param {AgreementArticleWithParent} article
  *
- * TODO Replace this dirty "let" by a clean recursive function.
+ * @returns {KaliData.AgreementArticleWithParentSections["sections"]}
  */
-function getArticleWithParentSections(articleId) {
-  const maybeIndexedArticle = INDEXED_ARTICLES.find(
-    indexedArticle => indexedArticle.articleId === articleId,
-  );
-  if (maybeIndexedArticle === undefined) {
-    throw new Error(`No agreement article found with this ID (${articleId}).`);
+const getSectionsFromArticle = article => {
+  const sections = [];
+  let section = article.parent;
+
+  while (section) {
+    const clonedSection = { ...section };
```

Comment:
not sure why you do that (expand + delete) but if you want a new object, you should clone section.data too as it will copy data ref into the cloned section